claim/gql_mutations.py

```diff
@@ -274,16 +276,23 @@ def create_attachments(claim_id, attachments):
         create_attachment(claim_id, attachment)
 
 
+@transaction.atomic
 def update_or_create_claim(data, user):
     items = data.pop('items') if 'items' in data else []
     services = data.pop('services') if 'services' in data else []
     incoming_code = data.get('code')
     claim_uuid = data.pop("uuid", None)
-    current_claim = Claim.objects.filter(uuid=claim_uuid).first()
-    current_code = current_claim.code if current_claim else None
-    if current_code != incoming_code \
-            and check_unique_claim_code(incoming_code):
-        raise ValidationError(_("mutation.code_name_duplicated"))
+    autogenerate_code = data.pop('autogenerate_code', None)
+    if autogenerate_code:
+        data['code'] = __autogenerate_claim_code()
+    else:
+        current_claim = Claim.objects.filter(uuid=claim_uuid).first()
+        current_code = current_claim.code if current_claim else None
+        if len(incoming_code) > ClaimConfig.max_claim_length:
+            raise ValidationError(_("mutation.code_name_too_long"))
+        if current_code != incoming_code \
+                and check_unique_claim_code(incoming_code):
+            raise ValidationError(_("mutation.code_name_duplicated"))
```

Comment:
1. If the code is autogenerated then it's uniqueness/format is not validated. I assume that code generated for Nepali has some date indicator that would assure uniqueness, but it could not be the case for other generators. Generated code should go through same validation as "regular" codes.
2. Should autogeneration be possible for update request as well?

Reply:
1. The autogenerating code is written in such a way that would guarantee its uniqueness (provided no additional entries have been added by hand ). It is inside atomic transaction, and generates new code based on the last used number. But since code field is not unique in the DB schema, if we assume that someone has access to DB and creates entries by hand, all validations became pointless in my opinion. If you change your system to autogeneration there is no possibility to create claim code by hand on the frontend.
2. No Idea about that, at the moment, I would ask to specify that requirement.

Reply:
1. Code uniqueness is true for the Nepali generator but we don't have control over generators that would be added later on. I believe that on claim level module uniqueness is a requirement, that currently could be unintentionally bypessed by autogenerator. Adding this check should make it more error-prone and doesn't require plenty of work. 
2. You've mentioned that when code is autogenerated then manual change of code is not possible. In that case i don't think any changes are required.